libqtile/widget/battery.py

```diff
@@ -267,7 +314,37 @@ def _get_param(self, name) -> tuple[str, str]:
 
         raise RuntimeError("Unable to read status for {}".format(name))
 
+    def set_battery_charge_thresholds(self, start, end):
+        if not self.charge_threshold_supported:
+            return
+
+        battery_dir = "/sys/class/power_supply"
+
+        path = os.path.join(battery_dir, self.battery, "charge_control_start_threshold")
+        try:
+            with open(path, "w+") as f:
+                f.write(str(start))
+        except FileNotFoundError:
+            self.charge_threshold_supported = False
+        except OSError:
+            logger.debug("Failed to write %s", path, exc_info=True)
+
+        path = os.path.join(battery_dir, self.battery, "charge_control_end_threshold")
+        try:
+            with open(path, "w+") as f:
+                f.write(str(end))
+        except FileNotFoundError:
+            self.charge_threshold_supported = False
+        except OSError:
+            logger.debug("Failed to write %s", path, exc_info=True)
+        return (start, end)
+
     def update_status(self) -> BatteryStatus:
+        (charge_start_threshold, charge_end_threshold) = self.charge_controller()
+        if self.force_charge:
+            charge_start_threshold = 0
+            charge_end_threshold = 100
+        self.set_battery_charge_thresholds(charge_start_threshold, charge_end_threshold)
```

Comment:
This will always set thresholds but what happens if a user has another script/app that's already setting charge limits? Should we have an option to disable smart charge completely?

Reply:
Yep, good point. Now you can set charge_controller to None and it shouldn't do anything.